src/optimize.ml

```diff
@@ -0,0 +1,123 @@
+open Types
+open Types.Simplified
+open Value
+
+let rec optimize_expr (exp : expr) : expr =
+  match exp with
+  | ((I32_const _ | I64_const _) as x) :: ((I32_const _ | I64_const _) as y) :: I_binop (nn, op) :: tl ->
+    begin try
+      let result = Interpret.exec_ibinop [(Value.of_instr y); (Value.of_instr x)] nn op in
+      begin
+        match result with
+        | [I32 _ | I64 _ as result ] -> optimize_expr (Value.to_instr result :: tl)
+        | _ -> assert false
+      end
+    with Trap _ -> x ::  optimize_expr (y :: I_binop (nn, op) :: tl)
+    end
+  | ((F32_const _ | F64_const _) as x) :: ((F32_const _ | F64_const _) as y) :: F_binop (nn, op) :: tl ->
+      let result = Interpret.exec_fbinop [(Value.of_instr y); (Value.of_instr x)] nn op in
+      begin
+        match result with
+        | [F32 _ | F64 _ as result ] -> optimize_expr (Value.to_instr result :: tl)
+        | _ -> assert false
+      end
+
+(* TODO ? another binary/unary ops *)
+
+  | (I32_const _ | I64_const _ | F32_const _ | F64_const _) :: Drop :: tl ->
+      optimize_expr tl
+  | Local_set x :: Local_get y :: tl when x = y ->
+      optimize_expr (Local_tee x :: tl)
+  | Br _ as br :: _tl -> [br]
+  (* | I32_const c :: (Br_if _ as brif) :: tl -> (* TODO : KO *)
+    begin match c with
+      | 0l -> optimize_expr tl
+      | _ -> [brif]
+    end *)
+  | Return :: _tl -> [Return]
+  | Return_call _ as rc :: _tl -> [rc]
+  | Return_call_indirect _ as rci :: _tl -> [rci]
+  | Nop :: tl -> optimize_expr tl
+  | Block (n, bt, e) :: tl ->
+    begin match optimize_expr e with
+    | [] -> optimize_expr tl
+    | oe -> Block (n, bt, oe) :: optimize_expr tl
+    end
+  | Loop (n, bt, e) :: tl ->
+    begin match optimize_expr e with
+    | [] -> optimize_expr tl
+    | oe -> Loop (n, bt, oe) :: optimize_expr tl
+    end
+  | If_else (n, bt, e1, e2) :: tl ->
+    begin match optimize_expr e1, optimize_expr e2 with
+    | [], [] -> Drop :: optimize_expr tl
+    | oe1, oe2 -> If_else (n, bt, oe1, oe2) :: optimize_expr tl
+    end
+  | hd :: tl -> hd :: optimize_expr tl
+  | [] -> []
+  
+let locals_func (body : expr) : (int, unit) Hashtbl.t =
+  let locals_hashtbl : (int, unit) Hashtbl.t = Hashtbl.create 10 in
+  let rec aux_instr (i : instr) : unit =
+    match i with
+    | Local_get ind | Local_set ind | Local_tee ind ->
+        Hashtbl.add locals_hashtbl ind ()
+    | Block (_, _, e) -> aux_expr e
+    | Loop (_, _, e) -> aux_expr e
+    | If_else (_, _, e1, e2) -> aux_expr e1; aux_expr e2
+    | _ -> ()
```

Comment:
This is going to be cumbersome, but it would probably be a good idea to list all cases here, so that when new cases are added in the future, we don't forget to go through them if it's needed.